In [1]:
from cmath import inf
from models import createFromBenchmark, Calculator

# Simple scheduling of job shop problems

Task 6. Design a data structure to store schedules.
Task 7. Design a class Calculator with a method to calculate the total operation time of a
schedule.
Task 8. Add a method to your calculator to generate the list all candidate schedules. Use
this method to calculate the makespan of a problem.
Task 9. Show experimentally why this approach is limited to small problems.

For this experiment we have used a quite small job shop problem. This is in order to easier visualize the solutions

In [2]:
filename = '/Users/ane/Projects/Performance Engineering/mini_project_4/benchmarks/benchmark_1.txt'
problem = createFromBenchmark(filename)
calc = Calculator()
print('Problem:')
print(problem)
print('Simple Schedule: ')
print(problem.getSchedule())

Problem:
Job 1 = [(3,4), (2,2), (1,1), ] /n Job 2 = [(1,3), (3,3), ] /n Job 3 = [(2,2), (1,4), (3,1), ] /n 
Simple Schedule: 
[(1, 0), (1, 1), (1, 2), (2, 0), (2, 1), (3, 0), (3, 1), (3, 2)]


In [3]:
#finding the best solution by checking all possible soloutions
schedules = calc.allCandidateSchedules(problem)
makespan = inf
best = None

for schedule in schedules:
    time = calc.totalOperationTime(problem, schedule)
    if time < makespan:
        makespan = time
        best = schedule

print('Best solution found: ', best)
print('Total operation time for the best solution: ', makespan)
for machine in problem.machines:
    s = 'Machine %s: %s' % (machine.id, machine.operations)
    print(s)

Length before removing:  40320
Length after removing:  560
Best solution found:  [(1, 0), (2, 0), (2, 1), (3, 0), (1, 1), (3, 1), (1, 2), (3, 2)]
Total operation time for the best solution:  8
Machine 1: [0, 0, 3, 3, 3, 3, 2, 2, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1]
Machine 2: [3, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0]
Machine 3: [0, 0, 0, 0, 0, 0, 3, 0, 0, 2, 2, 2, 1, 1, 1, 1, 0, 0, 0]


This solution is limited to small problems because it creates every possible combination of schedules and then removes the schedules that doesn't fit the requirements (Operations of a job must be performed in order, i.e. the next operation cannot be started
before the current is completed). As we can see from this small problem with only 8 operations in total, there are 40320 different schedules. This means that the isPossibleSchedule() function is ran over 40000 times. Then there are 560 possible schedules that we must compute the total timespan for. It is clearly that for a bigger problem this will take way too much time.

The permutations() function that we use for finding every possible combination actually terminates when the problem is too bi.